### PR TITLE
Add a bitwise NOT operator for integer and Boolean types (CPU).

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22326 Delegate Python ~ (invert operator) to Tensor.bitwise_not().
* #22320 Add a bitwise NOT operator for integer and Boolean types (CUDA).
* **#22283 Add a bitwise NOT operator for integer and Boolean types (CPU).**

Related to #20024, #22246, #22262

Differential Revision: [D16183576](https://our.internmc.facebook.com/intern/diff/D16183576)

### CODE REVIEW

**Comment by xuhdev** [comment]
In subsequent PRs, I will implement a GPU version and replace the `~` operator implementation with this function. But this PR by itself can also be seen as an independent contribution.

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 1]
```diff
@@ -51,6 +51,28 @@ static void abs_kernel(TensorIterator& iter) {
   });
 }
 
+// Boolean type does not work with ~ (bitwise NOT) in C++. bitwise_not_int_bool wraps this operation for both Boolean
```
Comment (izdeby) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp]: what is bitwise_not_int_bool?

File: test/test_torch.py [hunk 2]
```diff
@@ -1748,6 +1748,25 @@ def _test_neg(self, cast):
     def test_neg(self):
         self._test_neg(self, lambda t: t)
 
+    def test_bitwise_not(self):
+        res = 0xffff - torch.arange(127, dtype=torch.int8)
+        for t in (torch.BoolTensor, torch.LongTensor, torch.IntTensor, torch.ShortTensor, torch.CharTensor):
```
Comment (izdeby) [inline test/test_torch.py]: should here be also torch.ByteTensor?

File: test/test_torch.py [hunk 3]
```diff
@@ -1748,6 +1748,25 @@ def _test_neg(self, cast):
...
+            if t == torch.BoolTensor:
+                a = torch.BoolTensor([True, False])
+                this_res = torch.BoolTensor([False, True])
```
Comment (izdeby) [inline test/test_torch.py]: nit: 'expectedRes' or similar instead of this_res?

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -704,6 +704,22 @@ def merge_dicts(*dicts):
     tensor([0.0000, 0.0000, 0.0000, 1.0000, 1.0000, 0.0000, 0.5000])
 """)
 
+add_docstr(torch.bitwise_not,
```
Comment (izdeby) [inline torch/_torch_docs.py]: doc for inplace version of bitwise_not?

**Review by izdeby — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 1 above)
Comment (xuhdev) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp]: Updated

File: test/test_torch.py (hunk 2 above)
Comment (xuhdev) [reply test/test_torch.py]: Yes. Added.

File: test/test_torch.py (hunk 3 above)
Comment (xuhdev) [reply test/test_torch.py]: I changed to `expected_this`. I meant "result in this iteration".

File: torch/_torch_docs.py (hunk 4 above)
Comment (xuhdev) [reply torch/_torch_docs.py]: Added.

**Review by izdeby — DISMISSED** [review]

**Comment by izdeby** [comment]
are you going to add support for CUDA?

**Comment by xuhdev** [comment]
@izdeby Yes

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -339,6 +339,18 @@
     CPU: _bincount_cpu
     CUDA: _bincount_cuda
 
+- func: bitwise_not(Tensor self) -> Tensor
+  variants: function, method
+
+- func: bitwise_not_(Tensor(a!) self) -> Tensor(a!)
+  variants: function, method
```
Comment (colesbury) [inline aten/src/ATen/native/native_functions.yaml]: The in-place version should be method-only

Comment (xuhdev) [reply aten/src/ATen/native/native_functions.yaml]: Hmm... There are several other in-place versions with `function` variant. Should they also be changed?

Comment (xuhdev) [reply aten/src/ATen/native/native_functions.yaml]: For example, `abs_`, `ceil_`

Comment (colesbury) [reply aten/src/ATen/native/native_functions.yaml]: Yes, they should probably be changed (but not as part of this PR)

cc @gchanan

Comment (xuhdev) [reply aten/src/ATen/native/native_functions.yaml]: Sure -- Just to confirm I need to make these changes. Updated.

**Review by colesbury — DISMISSED** [review]
lgtm too

File: test/test_torch.py [hunk 6]
```diff
@@ -1748,6 +1748,40 @@ def _test_neg(self, cast):
...
+            # out
+            b = t()
+            torch.bitwise_not(a, out=b)
+            self.assertEqual(expected_res, b)
+            # in-place
+            a.bitwise_not_()
+            self.assertEqual(expected_res, a)
+
+        # test exceptions
+        for t in(torch.HalfTensor, torch.FloatTensor, torch.DoubleTensor):
```
Comment (xuhdev) [inline test/test_torch.py]: @izdeby I added more tests for exceptions here

**Comment by xuhdev** [comment]
@colesbury @izdeby I changed the code to avoid the template `bitwise_not` function (and thus the code becomes simpler)

**Comment by zou3519** [comment]
@izdeby @colesbury could you take another look at this please?

File: test/test_torch.py [hunk 7]
```diff
@@ -1748,6 +1748,40 @@ def _test_neg(self, cast):
     def test_neg(self):
         self._test_neg(self, lambda t: t)
 
+    def test_bitwise_not(self):
+        res = 0xffff - torch.arange(127, dtype=torch.int8)
+        for t in (torch.BoolTensor,
```
Comment (izdeby) [inline test/test_torch.py]: nit: can you, please, change variable names to something more readable? for example, 'dtype' instead of 't' and so on.

**Review by izdeby — APPROVED** [review]

File: test/test_torch.py (hunk 7 above)
Comment (xuhdev) [reply test/test_torch.py]: Yes. I made the corresponding changes in #22320 and #22326 as those are the places where dtype is used.

**Pull request closed without merging**